discovery.md

```diff
@@ -0,0 +1,436 @@
+# CloudSubscriptions: Discovery - Version 0.1-rc01
+
+## Abstract
+
+CloudSubscriptions Discovery API is a vendor-neutral API specification for
+determining what events an Event Producer or Aggregator / Broker
+has available, as well as how to subscribe to those events.
+
+## Status of this document
+
+This is a working draft. Breaking changes could be made in the next minor
+version.
+
+## Table of Contents
+
+- [Overview](#overview)
+- [Notations and Terminology](#notations-and-terminology)
+- [API Specification](#api-specification)
+- [Protocol Bindings](#protocol-bindings)
+- [Privacy & Security](#privacy-and-security)
+
+## Overview
+
+The goal of the CloudDiscovery API is to enable connection between consumers of
+events and producers / aggregators / brokers of events and to facilitate the
+creation of CloudSubscriptions.
+
+Discovery must allow for an event producer or intermediary component to
+advertise the event types that are available, provide the necessary information
+to consume that event (schema / delivery protocol options), and the necessary
+information to create a subscription. The output of the API must be such that
+tooling can be built where all possible event producers and types aren’t known
+in advance.
+
+There are several discovery use cases to consider from the viewpoint of event
+consumers.
+
+1. What sources are available, and what event types do they produce, and
+   (optionally) what `source` attributes are available?
+2. What event types are available, and from which providers.
+
+The second case becomes relevant if multiple providers support the same event
+types. Use case 1 is likely the dominant use case. Given the example of a public
+cloud provider where all producers produce events, there may be dozens of
+sources (producer systems) and hundreds of event types. The discovery funnel of
+producer first, then event types helps users navigate without having to see large
+lists of event types. Both of these cases require that filters can be used
+in the discovery API to narrow down the selection of available events.
+
+The CloudEvent source attribute is a potential source of high fanout. For
+example consider a blob storage system where each directory constitutes a
+distinct `source` attribute value. For this reason, source is not the starting
+point for discovery, but an attribute that can be used when creating
+subscriptions. A producer is free to show different possible source values
+to different users in their system.
+
+There are use cases where the Discovery API is served as a collection of other
+producers. These take two forms:
+
+1. Aggregator
+2. Events Broker
+
+In the aggregator case, discovery is a combined set of producers for where the
+discovery is passed through largely unmodified and subscriber URIs are generally
+absolute (back to the producer).
+
+In the events broker use case, it is allowable to modify the output of
+discovery and subscription URIs will be relative to the broker itself.
```

Comment:
Do you think the consumer will need to know when each of these two cases is actually in use? I was kind of assuming they wouldn't, and if that's the case then it might be good to add a sentence at the end mentioning that so people understand that these two cases are just mentioned to help explain the use cases for the spec and not because it'll impact consumer implementation details.

Reply:
In trying to add wording or rewrite this, I decided it is best removed.